cdc/kv/region_worker.go

```diff
@@ -0,0 +1,390 @@
+	metricPullEventCommittedCounter := pullEventCounter.WithLabelValues(cdcpb.Event_COMMITTED.String(), captureAddr, changefeedID)
+	metricPullEventCommitCounter := pullEventCounter.WithLabelValues(cdcpb.Event_COMMIT.String(), captureAddr, changefeedID)
+	metricPullEventPrewriteCounter := pullEventCounter.WithLabelValues(cdcpb.Event_PREWRITE.String(), captureAddr, changefeedID)
+	metricPullEventRollbackCounter := pullEventCounter.WithLabelValues(cdcpb.Event_ROLLBACK.String(), captureAddr, changefeedID)
+	metricSendEventResolvedCounter := sendEventCounter.WithLabelValues("native-resolved", captureAddr, changefeedID)
+	metricSendEventCommitCounter := sendEventCounter.WithLabelValues("commit", captureAddr, changefeedID)
+	metricSendEventCommittedCounter := sendEventCounter.WithLabelValues("committed", captureAddr, changefeedID)
+
+	var err error
+	for {
+		select {
+		case <-ctx.Done():
+			return errors.Trace(ctx.Err())
+		case event, ok := <-w.inputCh:
+			if !ok {
+				log.Debug("region worker receiver closed")
+				return nil
+			}
+			// event == nil means the region worker should exit and re-establish
+			// all existing regions.
+			if event == nil {
+				log.Info("region worker closed by error")
+				return w.evictAllRegions(ctx)
+			}
+			if event.state.isStopped() {
+				continue
+			}
+			event.state.lock.Lock()
+			if event.changeEvent != nil {
+				metricEventSize.Observe(float64(event.changeEvent.Event.Size()))
+				switch x := event.changeEvent.Event.(type) {
+				case *cdcpb.Event_Entries_:
+					err = w.handleEventEntry(
+						ctx, x, event.state,
+						metricPullEventInitializedCounter,
+						metricPullEventPrewriteCounter,
+						metricPullEventCommitCounter,
+						metricPullEventCommittedCounter,
+						metricPullEventRollbackCounter,
+						metricSendEventCommitCounter,
+						metricSendEventCommittedCounter,
+					)
+					if err != nil {
+						err = w.handleSingleRegionError(ctx, err, event.state)
+					}
+				case *cdcpb.Event_Admin_:
+					log.Info("receive admin event", zap.Stringer("event", event.changeEvent))
+				case *cdcpb.Event_Error:
+					err = w.handleSingleRegionError(
+						ctx,
+						cerror.WrapError(cerror.ErrEventFeedEventError, &eventError{err: x.Error}),
+						event.state,
+					)
+				case *cdcpb.Event_ResolvedTs:
+					if err = w.handleResolvedTs(ctx, x.ResolvedTs, event.state, metricSendEventResolvedCounter); err != nil {
+						err = w.handleSingleRegionError(ctx, err, event.state)
+					}
+				}
+			}
+
+			if event.resolvedTs != nil {
+				if err = w.handleResolvedTs(ctx, event.resolvedTs.Ts, event.state, metricSendEventResolvedCounter); err != nil {
+					err = w.handleSingleRegionError(ctx, err, event.state)
+				}
+			}
+			event.state.lock.Unlock()
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (w *regionWorker) run(ctx context.Context) error {
+	wg, ctx := errgroup.WithContext(ctx)
+	wg.Go(func() error {
+		return w.resolveLock(ctx)
```

Comment:
Could regions share few resolvedLock goroutines?

Reply:
I don't quite understand your meaning

Reply:
current design, is one gRPC stream <-> one worker (multiple routines support in another pr) <-> one resolve lock routine